File: Kodlar/Soru 3/GAPasswordCracker/src/Fitness.java
/*
 * EGE DOĞAN DURSUN - 05170000006
 * CEM ÇORBACIOĞLU - 05130000242
 * EGE ÜNİVERSİTESİ
 * MÜHENDİSLİK FAKÜLTESİ
 * BİLGİSAYAR MÜHENDİSLİĞİ BÖLÜMÜ
 * 2019-2020 BAHAR DÖNEMİ
 * İŞLEMSEL ZEKA VE DERİN ÖĞRENME DERSİ - CIDL2020-P1
 * SORU 3 - GENETİK ALGORİTMAYLA ŞİFRE ÇÖZME
 * TARİH : 12 NİSAN 2020
 * 
 */

import java.util.ArrayList;

public class Fitness {

	
	//This function evaluates the fitness level of the population. Max possible fitness level is
	//the maximum character amount of the target string.
	public static ArrayList<Individual> evaluateFitness(ArrayList<Individual> population, String target) {
		
		for (int i = 0; i<population.size(); i++) {
			
			int fitness = 0;
			Individual curIndividual = population.get(i);
			for (int j = 0; j<curIndividual.getChromosomeLength(); j++) {
				if (target.charAt(j) == curIndividual.getChromosome().charAt(j)) {
					fitness = fitness + 1;
				}
			}
			curIndividual.setFitness(fitness);
		}
		return population;
	}
	
	//This function shows the average fitness level of a population based on averaging 
	//how many characters are matching with the target string.
	public static double showAverageFitness(ArrayList<Individual> population) {
		
		double summation = 0;
		for (int i = 0; i<population.size(); i++) {
			summation = summation + population.get(i).getFitness();
		}
		double average = summation/population.size();
		
		return average;
	}
	
	
	//This function returns exactly how many members of the population provides a perfect
	//100% match with the target string to provide insight.
	public static int calculateCorrectMatches(ArrayList<Individual> population, String target) {
		
		int total = 0;
		for(int i = 0; i<population.size(); i++) {
			if(population.get(i).getChromosome().equals(target)) {
				total = total + 1;
			}
		}
		
		return total;
	}
}

File: Kodlar/Soru 3/GATravellingSalesmanProblem/Fitness.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
EGE DOĞAN DURSUN - 05170000006
CEM ÇORBACIOĞLU - 05130000242
EGE ÜNİVERSİTESİ
MÜHENDİSLİK FAKÜLTESİ
BİLGİSAYAR MÜHENDİSLİĞİ BÖLÜMÜ
2019-2020 BAHAR DÖNEMİ
İŞLEMSEL ZEKA VE DERİN ÖĞRENME DERSİ  - CIDL2020-P1
PROJE 1 - SORU 3 - ALTERNATİF 3: GENETİK ALGORİTMALARLA TRAVELLING SALESMAN PROBLEMİ ÇÖZÜMÜ
TARİH : 13 NİSAN 2020
"""

class Fitness:
    def __init__(self, route):
        self.route = route
        self.distance = 0
        self.fitness= 0.0
    
    #Method to calculate the distance for the route
    def route_distance(self):
        if self.distance ==0:
            path_distance = 0
            for i in range(0, len(self.route)):
                from_city = self.route[i]
                to_city = None
                
                if i + 1 < len(self.route):
                    to_city = self.route[i + 1]
                else:
                    to_city = self.route[0]
                    
                path_distance = path_distance + from_city.distance(to_city)
            self.distance = path_distance
        return self.distance
    
    #Method to calculate the route fitness depending on the route distance
    def route_fitness(self):
        if self.fitness == 0:
            self.fitness = 1 / float(self.route_distance())
        return self.fitness
    
    